src/soroban_rpc.ts

```diff
@@ -162,16 +162,77 @@ export namespace SorobanRpc {
     retval: xdr.ScVal;
   }
 
-  export interface SimulateTransactionResponse {
+  /**
+   * Simplifies {@link RawSimulateTransactionResponse} into separate interfaces
+   * based on status:
+   *   - on success, this includes all fields, though `result` is only present
+   *     if an invocation was simulated (since otherwise there's nothing to
+   *     "resultify")
+   *   - if there was an expiration error, this includes error and restoration
+   *     fields
+   *   - for all other errors, this only includes error fields
+   *
+   * @see https://soroban.stellar.org/api/methods/simulateTransaction#returns
+   */
+  export type SimulateTransactionResponse =
+    | SimulateTransactionSuccessResponse
```

Comment:
great design, how do you feel about sub-classing `SimulateSuccessResponse` a bit further to capture those other conditional data models so clients aren't (re)implementing same on their side?
```
SimulateInvokeHostFnResponse extends SimulateSuccessResponse
SimulateExpiredResponse extends SimulateSuccessResponse
```

`SimulateSuccessResponse` still used as catch-all on remainder of success sim responses that have no further differentiation.

Reply:
I like the intention, but TypeScript has a **very** annoying limitation in that you can't just do 

```typescript
if (response instanceof SimulateExpiredResponse) { /* ... handle it */ }
```

like you can in other languages, since interfaces are basically only for transpile-time type checking. So even if we added these, users would **still** need to do the equivalent of

```typescript
if (isSimulateExpiredResponse(response)) { /* ... handle it */ }
```

whose implementation (as you can see [here](https://github.com/stellar/js-soroban-client/pull/132/files#diff-fff0ee1c855d7fdef5ec027268d27066a50d0c0dc4e4832b51f8a2e40129e177R192-R198)) is just a bunch of fragile "does it have this property" conditionals.

Plus, clients would have to have way more checks to "break down" the union type (i.e. lots more like [this diff](https://github.com/stellar/js-soroban-client/pull/132/files#diff-fff0ee1c855d7fdef5ec027268d27066a50d0c0dc4e4832b51f8a2e40129e177R55-R57) in the PR). (Then again, you could just `as <ExpectedType>` it, but that's also annoying / fragile.) I think it may be, for now, a little more user-friendly to do

```typescript
if (!!response.restorePreamble) { /* ... handle it */ }
```

as it explicitly marks the field that the client can expect to be present.

My rationale is primarily motivated by the use-case here: I think that users know which fields to expect in which cases. For example, if you **know** you're simulating an invocation, you _also_ know that there will be a `result` field in the success case. Then, you _also_ know that you should check for the `restorePreamble` field prior to submission. So there is less of a need to break down the success union type.

I guess if we had more response variations like this in the future (e.g. more fields or branches), I would be more in support of the interface breakdown. But as such, it might be types for the sake of types at the cost of unnecessary client-side code.

Reply:
Actually, I added `isSimulationRestore` in a8e2001 to break it down at least a bit further.